### PR TITLE
HOTFIX: fix subscription narrowing problem

### PR DESCRIPTION
**Pull request by vejrj**

### CODE REVIEW

File: packages/ts-codegen/src/__tests__/index.test.ts [hunk 1]
```diff
@@ -179,10 +179,14 @@ describe(generateTS, () => {
                 readonly userUpdated?: userUpdated<any>;
             }
             export type userUpdated<SubscribeResult = never> = {
+                /*We added this to fix the typescript narrowing problem.*/
+                kind?: "Subscribe";
                 subscribe: (model: unknown, args: {}, context: unknown, info: ResolveInfo) => PromiseOrValue<AsyncIterator<{
```
Comment (alloy) [inline packages/ts-codegen/src/__tests__/index.test.ts:184]: These comments seem unnecessary, just have the commit/PR have a good description of the problem.

**Review by alloy — COMMENTED** [review]
I’m missing an example of what the resolver implementation would look like. Is this requiring folks to include the kind property in their implementation?

**Comment by vejrj** [comment]
> I’m missing an example of what the resolver implementation would look like. Is this requiring folks to include the kind property in their implementation?

Yes, you will need to add kind for "SubscribeAndResolve"

**Comment by alloy** [comment]
That doesn’t seem helpful. At that point it’s more intuitive to export two different type definitions.

**Comment by vejrj** [comment]
> That doesn’t seem helpful. At that point it’s more intuitive to export two different type definitions.

That was my initial idea, but @freiksenet said that we would need to refactor everything then.

**Comment by alloy** [comment]
Can you elaborate on the difference in effort needed to add `kind` vs using a different type? Both can be done relatively the same way with a codemod afaik.

**Review by alloy — CHANGES_REQUESTED** [review]
As discussed offline, I consider this solution a no-go. We provide framework solutions and we need to consider our customers [the product engineers], and this workaround is not intuitive. We should:

1. Figure out if it's simply not possible to provide a single type and have TypeScript correctly narrow the union;
2. and otherwise provide 2 types and have the product engineer choose the right one—which is something that they can intuitively understand.

**Pull request closed without merging**